src/printer_tree.c

```diff
@@ -0,0 +1,2965 @@
+struct trt_options
+{
+    uint32_t max_linebreaks;    /**< Max linebreaks per section. Functionality is not implemented. */
+    uint32_t *cnt_linebreak;    /**< Pointer to struct trt_printing.cnt_linebreak counter. Functionality is not implemented. */
+};
+
+/**
+ * @brief Main structure for browsing the libyang tree
+ */
+struct trt_tree_ctx
+{
+    trt_actual_section section;         /**< To which section pn points. */
+    int64_t index_within_section;       /**< For example in which 'grouping' we are right now. It has the number -1 if it is invalid. */
+    int64_t index;                      /**< Some nodes contain an array of other nodes etc. list has actions. Number -1 means invalid value. */
+    const struct lys_module *module;    /**< Schema tree structures. */
+    const struct lysp_node *pn;         /**< Actual pointer to parsed node. */
+    const struct lysp_node *tpn;        /**< Pointer to actual top-node. */
+    struct trt_options opt;             /**< Options for printing. */
+};
+
+/** Getter function for tro_lysp_node_charptr function. */
+typedef const char *(*trt_get_charptr_func)(const struct lysp_node *pn);
+
+/******************************************************************************
+ * Declarations end
+ *****************************************************************************/
+/******************************************************************************
+ * Definitions start
+ *****************************************************************************/
+/******************************************************************************
+ * Definition of the general Trg / Trp functions
+ *****************************************************************************/
+
+/** Set .cnt_linebreak to zero. */
+static void
+trp_cnt_linebreak_reset(struct trt_printing *p)
+{
+    p->cnt_linebreak = 0;
+}
+
+/** Increment .cnt_linebreak by one. */
+static void
+trp_cnt_linebreak_increment(struct trt_printing *p)
+{
+    p->cnt_linebreak++;
+}
+
+/**
+ * @brief Print function.
+ *
+ * This is a special abstraction for printing characters on output.
+ * It was created because in some places in the code you only need to know the number of characters to print
+ * in which case nothing may be printed.
+ * So just set the callback function for counting characters (trt_printing.pf)
+ * and nothing else in the code needs to change, because other functions do not know the difference.
+ * Variadic arguments are expected to be of type char*.
+ *
+ * @param[in] p is used for printing or counting characters.
+ * @param[in] arg_count number of arguments in va_list.
+ */
+static void
+trp_print(struct trt_printing *p, int arg_count, ...)
+{
+    va_list ap;
+
+    va_start(ap, arg_count);
+    p->pf(p->out, arg_count, ap);
+    va_end(ap);
+}
+
+#define PRINT_N_TIMES_BUFFER_SIZE 16
+
+/** Print character n times. */
+static void
+trg_print_n_times(int32_t n, char c, struct trt_printing *p)
```

Comment:
if using `ly_print()`, which wrapps `printf()`, the `"%*c", n, c` should do the job